### PR TITLE
Add 'identified' and 'managed' node event

### PR DESCRIPTION
**Pull request by anhou**
For ORFS 246/247  docs: https://github.com/RackHD/docs/pull/296  'discovered' and 'removed' related PRs have been merged in previous PRs

## NOTE: WIP

@RackHD/corecommitters @RackHD/rackhd_dev

### CODE REVIEW

File: lib/models/node.js [hunk 1]
```diff
@@ -194,6 +204,31 @@ function NodeModelFactory (Model, waterline, _, Promise, Constants, configuratio
                 return next();
             }
             return next();
+        },
+
+        publishEvent: function(node, next) {
+            return waterline.obms.find({node: node.id})
+            .then(function(obms) {
+                if(node.sku && _.isEmpty(obms)) {
```
Comment (WangWinson) [inline lib/models/node.js]: Will below be a little easy ?

``` javascript
if (node.sku) {
    return eventsProtocol.publishNodeAlert(node.id, {
          nodeId : node.id,
          nodeType: node.type,
          state: _.isEmpty(obms) ? "identified" : "managed"
    });
}
```

Comment (yyscamper) [inline lib/models/node.js]: Agree

File: lib/models/node.js [hunk 2]
```diff
@@ -194,6 +204,31 @@ function NodeModelFactory (Model, waterline, _, Promise, Constants, configuratio
...
+                }
+            }).then(function() {
+                next();
+            }).catch(function(err) {
+                next(err);
+            });
+        },
+
+        afterUpdate: function (node, next) {
+            this.publishEvent(node, next);
```
Comment (yyscamper) [inline lib/models/node.js]: What if only the node's name or identifier or other fields (these not either sku or obm) are changed? I think there is no need to publish event is such cases.

Comment (anhou) [inline lib/models/node.js]: good catch, updated

File: lib/models/node.js (hunk 1 above)
Comment (anhou) [inline lib/models/node.js]: @WangWinson Good,  didn't pay attention, updated

File: lib/models/obms.js [hunk 3]
```diff
@@ -106,6 +108,37 @@ function ObmsModelFactory (
...
+        publishEvent: function(nodeId, next) {
+            return waterline.nodes.findByIdentifier(nodeId)
+            .then(function(node) {
+                if (!_.isEmpty(node)) {
+                    return waterline.obms.find({node: node.id})
+                    .then(function(obms) {
+                        return eventsProtocol.publishNodeAlert(node.id,
+                            { nodeId : node.id,
+                              nodeType: node.type,
+                              state: _.isEmpty(obms) ? 'identified': 'managed' });
```
Comment (zyoung51) [inline lib/models/obms.js]: Do you want to check for the node.sku here to decide if it is 'identified'?

File: lib/models/obms.js [hunk 4]
```diff
@@ -106,6 +108,37 @@ function ObmsModelFactory (
             next();
         },
 
+        publishEvent: function(nodeId, next) {
+            return waterline.nodes.findByIdentifier(nodeId)
```
Comment (zyoung51) [inline lib/models/obms.js]: Would 'needByIdentifier' be useful to clean up the logic below?

File: lib/models/obms.js [hunk 5]
```diff
@@ -106,6 +108,37 @@ function ObmsModelFactory (
...
+            .then(function(node) {
+                if (!_.isEmpty(node)) {
+                    return waterline.obms.find({node: node.id})
```
Comment (zyoung51) [inline lib/models/obms.js]: If 'created/updated' is calling this can you skip going back to the database and assert that it is 'managed'.

**Comment by zyoung51** [comment]
How does the addition of these AMQP notifications and extra database accesses affect the latency of database operations?  Can these notifications occur asynchronously without waiting for the promise to be resolved?  Is the inability to emit an event a good reason to error out the database operation?

File: lib/models/obms.js (hunk 3 above)
Comment (anhou) [inline lib/models/obms.js]: WIP PR, many logic holes here

File: lib/models/obms.js (hunk 4 above)
Comment (anhou) [inline lib/models/obms.js]: needByIdnetifier will throw Errors.NotFoundError,  not intended to let it throw error

**Comment by anhou** [comment]
@zyoung51 Thanks for comments, it's a WIP PR to try to attempt to achieve publish node events in models,  the comments you address such as latency/performance, error handling is good concerns, and more, I found that some previous record information for stage change cannot be got from lifecycle functions, so I'll close this PR and create other PRs for 'identified' and 'managed'  events

File: lib/models/obms.js (hunk 4 above)
Comment (zyoung51) [inline lib/models/obms.js]: Is your thought that the .catch() on line 125 would not be able to handle an exception thrown in the promise chain?

**Pull request closed without merging**